File: backend/agents/instagram_publisher.py
import os
import time
import base64
import requests
from config import (
    INSTAGRAM_ACCESS_TOKEN, INSTAGRAM_ACCOUNT_ID,
    INSTAGRAM_BASE_URL, IMGBB_API_KEY
)


class InstagramPublisher:

    def publicar_tudo(self, criativos: dict, caption: str, video_url: str = None) -> dict:
        resultado = {}

        # Feed / Carrossel
        if criativos.get("carousel"):
            try:
                post_id = self.publicar_carrossel(criativos["carousel"], caption)
                resultado["feed"] = {"sucesso": True, "id": post_id}
            except Exception as e:
                resultado["feed"] = {"sucesso": False, "erro": str(e)}

        # Stories
        if criativos.get("stories"):
            try:
                stories_id = self.publicar_stories(criativos["stories"][0])
                resultado["stories"] = {"sucesso": True, "id": stories_id}
            except Exception as e:
                resultado["stories"] = {"sucesso": False, "erro": str(e)}

        # Reels (apenas se tiver vídeo)
        if video_url:
            try:
                reels_id = self.publicar_reels(video_url, caption)
                resultado["reels"] = {"sucesso": True, "id": reels_id}
            except Exception as e:
                resultado["reels"] = {"sucesso": False, "erro": str(e)}

        return resultado

    def publicar_carrossel(self, caminhos_imagens: list[str], caption: str) -> str:
        if len(caminhos_imagens) == 1:
            return self._publicar_imagem_unica(caminhos_imagens[0], caption)

        urls = [self._hospedar_imagem(p) for p in caminhos_imagens[:10]]

        containers = []
        for url in urls:
            r = requests.post(
                f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media",
                params={
                    "image_url": url,
                    "is_carousel_item": "true",
                    "access_token": INSTAGRAM_ACCESS_TOKEN,
                }
            )
            self._checar_erro(r)
            containers.append(r.json()["id"])
            time.sleep(1)

        r = requests.post(
            f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media",
            params={
                "media_type": "CAROUSEL",
                "children": ",".join(containers),
                "caption": caption,
                "access_token": INSTAGRAM_ACCESS_TOKEN,
            }
        )
        self._checar_erro(r)
        carousel_id = r.json()["id"]

        return self._publicar_container(carousel_id)

    def publicar_stories(self, caminho_imagem: str) -> str:
        url = self._hospedar_imagem(caminho_imagem)

        r = requests.post(
            f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media",
            params={
                "image_url": url,
                "media_type": "STORIES",
                "access_token": INSTAGRAM_ACCESS_TOKEN,
            }
        )
        self._checar_erro(r)
        container_id = r.json()["id"]

        return self._publicar_container(container_id)

    def publicar_reels(self, video_url: str, caption: str) -> str:
        r = requests.post(
            f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media",
            params={
                "video_url": video_url,
                "media_type": "REELS",
                "caption": caption,
                "access_token": INSTAGRAM_ACCESS_TOKEN,
            }
        )
        self._checar_erro(r)
        container_id = r.json()["id"]

        # Aguarda processamento do vídeo
        for _ in range(24):
            time.sleep(5)
            r = requests.get(
                f"{INSTAGRAM_BASE_URL}/{container_id}",
                params={"fields": "status_code", "access_token": INSTAGRAM_ACCESS_TOKEN}
            )
            status = r.json().get("status_code")
            if status == "FINISHED":
                break
            elif status == "ERROR":
                raise Exception("Erro no processamento do vídeo no Instagram.")

        return self._publicar_container(container_id)

    def _publicar_imagem_unica(self, caminho: str, caption: str) -> str:
        url = self._hospedar_imagem(caminho)
        r = requests.post(
            f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media",
            params={
                "image_url": url,
                "caption": caption,
                "access_token": INSTAGRAM_ACCESS_TOKEN,
            }
        )
        self._checar_erro(r)
        container_id = r.json()["id"]
        return self._publicar_container(container_id)

    def _publicar_container(self, container_id: str) -> str:
        r = requests.post(
            f"{INSTAGRAM_BASE_URL}/{INSTAGRAM_ACCOUNT_ID}/media_publish",
            params={
                "creation_id": container_id,
                "access_token": INSTAGRAM_ACCESS_TOKEN,
            }
        )
        self._checar_erro(r)
        return r.json()["id"]

    def _hospedar_imagem(self, caminho: str) -> str:
        if not IMGBB_API_KEY:
            raise ValueError("IMGBB_API_KEY não configurada no .env")

        with open(caminho, "rb") as f:
            imagem_b64 = base64.b64encode(f.read()).decode("utf-8")

        r = requests.post(
            "https://api.imgbb.com/1/upload",
            data={"key": IMGBB_API_KEY, "image": imagem_b64},
            timeout=30
        )
        r.raise_for_status()
        return r.json()["data"]["url"]

    def _checar_erro(self, response: requests.Response):
        if not response.ok:
            try:
                erro = response.json().get("error", {}).get("message", response.text)
            except Exception:
                erro = response.text
            raise Exception(f"Instagram API erro {response.status_code}: {erro}")

File: lib/instagram-publisher.ts
import { prepareInstagramImages } from "@/lib/instagram-image";
import { ensureInstagramQuota, recordInstagramPublication } from "@/lib/publications";
import { readTenantSettings } from "@/lib/settings";

const graphVersion = process.env.META_GRAPH_VERSION || "v25.0";
const graphBase = `https://graph.facebook.com/${graphVersion}`;
const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function graphFetch(url: string, init?: RequestInit) {
  const response = await fetch(url, {
    ...init,
    signal: AbortSignal.timeout(15000)
  });
  const data = await response.json();
  if (!response.ok) {
    throw new Error(data?.error?.message || `Instagram API retornou ${response.status}.`);
  }
  return data;
}

async function fetchPublishedPermalink(mediaId: string, token: string) {
  if (!mediaId) return "";
  try {
    const data = await graphFetch(
      `${graphBase}/${encodeURIComponent(mediaId)}?fields=permalink&access_token=${encodeURIComponent(token)}`
    );
    return String(data?.permalink || "");
  } catch {
    return "";
  }
}

async function publishToInstagram(caption: string, imageUrls: string[], accountId: string, token: string) {
  if (!accountId || !token) throw new Error("Credenciais do Instagram nao configuradas.");

  const create = async (imageUrl: string, isCarouselItem: boolean) => {
    const params = new URLSearchParams({ image_url: imageUrl, access_token: token });
    if (isCarouselItem) params.set("is_carousel_item", "true");
    else params.set("caption", caption);
    const data = await graphFetch(`${graphBase}/${accountId}/media`, { method: "POST", body: params });
    return String(data.id);
  };

  const waitUntilReady = async (creationId: string) => {
    let lastStatus = "";
    for (let attempt = 0; attempt < 18; attempt++) {
      const data = await graphFetch(
        `${graphBase}/${creationId}?fields=status_code&access_token=${encodeURIComponent(token)}`
      );
      lastStatus = String(data.status_code || "");
      if (lastStatus === "FINISHED") return;
      if (lastStatus === "ERROR") throw new Error("Instagram retornou erro ao processar uma imagem do carrossel.");
      await sleep(2000);
    }
    throw new Error(`Instagram ainda nao liberou a midia para publicacao. Status: ${lastStatus || "pendente"}.`);
  };

  const publish = async (creationId: string) => {
    await waitUntilReady(creationId);
    return graphFetch(`${graphBase}/${accountId}/media_publish`, {
      method: "POST",
      body: new URLSearchParams({ creation_id: creationId, access_token: token })
    });
  };

  if (imageUrls.length === 1) {
    const result = await publish(await create(imageUrls[0], false));
    return { ...result, tipo: "imagem", fotos_publicadas: 1 };
  }

  const children = await Promise.all(imageUrls.slice(0, 10).map((imageUrl) => create(imageUrl, true)));
  await Promise.all(children.map((childId) => waitUntilReady(childId)));

  const carousel = await graphFetch(`${graphBase}/${accountId}/media`, {
    method: "POST",
    body: new URLSearchParams({
      media_type: "CAROUSEL",
      children: children.join(","),
      caption,
      access_token: token
    })
  });
  const result = await publish(String(carousel.id));
  return { ...result, tipo: "carrossel", fotos_publicadas: children.length };
}

export async function publishInstagramForTenant(input: {
  tenantId: string;
  userId?: string | null;
  propertyCode?: string | number | null;
  caption: string;
  imageUrls: string[];
}) {
  const settings = await readTenantSettings(input.tenantId);
  const caption = String(input.caption || "").trim();
  const imageUrls = Array.from(new Set(input.imageUrls || [])).filter((url) => /^https?:\/\//i.test(url));

  if (!caption) throw new Error("A legenda do Instagram nao pode ficar vazia.");
  if (!imageUrls.length) throw new Error("Selecione pelo menos uma imagem publica para o Instagram.");

  await ensureInstagramQuota(input.tenantId);
  const preparedImageUrls = await prepareInstagramImages(imageUrls.slice(0, 10), {
    tenantId: input.tenantId,
    preset: "feed-portrait"
  });
  const result = await publishToInstagram(caption, preparedImageUrls, settings.instagramAccountId, settings.instagramAccessToken || "");
  const publishedPermalink = await fetchPublishedPermalink(String(result.id || ""), settings.instagramAccessToken || "");
  const permalink =
    publishedPermalink || result.permalink || result.url || (result.id ? `https://www.instagram.com/p/${result.id}/` : "");
  const { usage } = await recordInstagramPublication({
    tenantId: input.tenantId,
    userId: input.userId,
    propertyCode: input.propertyCode,
    caption,
    instagramPostId: result.id,
    instagramUrl: permalink,
    mediaType: result.tipo,
    photosCount: result.fotos_publicadas,
    metadata: { ...result, originalImageUrls: imageUrls, preparedImageUrls, imagePreset: "feed-portrait" }
  });
  return {
    success: true,
    provider: "instagram" as const,
    postId: String(result.id || ""),
    url: permalink,
    result: { ...result, url: permalink, imagens_tratadas: preparedImageUrls.length },
    usage
  };
}
